### PR TITLE
✨ Source Salesforce: Bulk stream uses async CDK components

### PR DESCRIPTION
**Pull request by maxi297**
## What
Following [the CDK release of the improve async job components](https://github.com/airbytehq/airbyte/pull/45673), we will be able to release Salesforce that relies on those CDK components in order to sync bulk streams.

## How
See https://github.com/airbytehq/airbyte/pull/45673 for more details.

We are currently rolling out this change using dev version `2.5.32-dev.0a58b0968c`. The goal for the progressive rollout is roughly:
* 2024-09-19: internal workspaces
* 2024-09-20: 5 external workspaces (at least one including parent streams) (see update [here](https://airbytehq-team.slack.com/archives/C02U9R3AF37/p1726840931768489?thread_ts=1726758006.091299&cid=C02U9R3AF37))
* 2024-09-23: 20 external workspaces
* 2024-09-24: Full release if no issues discovered

## Review guide
See https://github.com/airbytehq/airbyte/pull/45673 for more details.

## User Impact
There should be no user impact as the goal is to make the maintenance easier for us. 

There is one case we willingly changed the behavior and it his [here](https://github.com/airbytehq/airbyte/pull/45678/files#diff-0572edeedd561901f8110c0f88b48efde87d1ed91247c0eb189c7229da12090dL647) where before, the connector would retry a whole job given we could not download the result. It has been removed because:
* It is harder to port back to the new version of the CDK because the job creation and the download are now in two separate code paths (stream_slices and read_records instead of just read_records)
* The value is very unclear as this was done as an attempted patch without being able to clearly monitor the value of the fix. ([We see the warning log in the last 7 days](https://console.cloud.google.com/logs/query;query=%22Downloading%20data%20failed%20after%22;summaryFields=:false:32:beginning;cursorTimestamp=2024-09-20T13:17:02.338977009Z;duration=P7D?project=prod-ab-cloud-proj&pli=1)). All of these happened on the same connection, `8bb4614b-cc29-41b0-bc7c-86d4f52bab62`. As there were  5 logs for `Downloading data failed after 0 retries. Retrying the whole job...`, there were only two for `Downloading data failed even after 1 retries. Stopping retry and raising exception` which seems to indicate that retrying helped in that specific case or the stream stopped before. As the logs are a bit screwed up, I can't determine which one is true here. In any case, I would assume that we can push back if it's only for one customer and retrying would work on the next attempt/job.

Given that we were to see that case in prod, the fix would be to have `AsyncRetriever.read_records` create a new factory using it's factory and having just one slice.

## Can this PR be safely reverted and rolled back?

- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-salesforce/pyproject.toml [hunk 1]
```diff
@@ -16,9 +16,9 @@ repository = "https://github.com/airbytehq/airbyte"
 include = "source_salesforce"
 
 [tool.poetry.dependencies]
-python = "^3.9,<3.12"
+python = "^3.10,<3.12"
 pandas = "2.2.1"
-airbyte-cdk = "^2"
+airbyte-cdk = "^5"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-salesforce/pyproject.toml:21]: Need to regenerate lock once CDK is released

**Comment by maxi297** [comment]
We are getting a seg fault on some of the workspaces we've progressively rolled out to. We have two pre-build that might help us debug:
* 2.6.0-dev.884dfdee72: Better handling on breaking errors during polling
* 2.6.0-dev.41de741141: Better logging on seg fault crash

File: airbyte-integrations/connectors/source-salesforce/unit_tests/test_memory.py [hunk 2]
```diff
@@ -1,46 +0,0 @@
...
-        (200000, 7, 19),
-    ),
-    ids=[
-        "1k recods",
-        "10k records",
-        "100k records",
-        "200k records",
-    ],
-)
-def test_memory_download_data(stream_config, stream_api, n_records, first_size, first_peak):
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/test_memory.py:29]: There was indeed a memory issue. I created [this PR](https://github.com/airbytehq/airbyte/pull/46286) to address the issue directly in the CDK. Once you approve it, I'll resolve this issue

**Review by artem1205 — APPROVED** [review]
LGTM!

**Pull request merged**